### PR TITLE
sql: fix bad "split"s in fakeSpanResolver

### LINKED ISSUE #44323: sql: fakeSpanResolver can "split" SQL rows across nodes
This issue affects logic tests only. For tables with multiple column families, `fakeSpanResolver` doesn't necessarily map spans corresponding to the same row to the same node. This can cause spurious logic test failures for `fakedist` configurations.

Example:
```
# LogicTest: fakedist

statement ok
CREATE TABLE t (a INT PRIMARY KEY, b INT, c INT NOT NULL, FAMILY (a), FAMILY (b), FAMILY (c))

statement ok
INSERT INTO t VALUES (1, 1, 1)

query I
SELECT c FROM t WHERE a = 1
----
1
```
This test sometimes passes but sometimes fails with
```
testdata/logic_test/__test:9: SELECT c FROM t WHERE a = 1
expected:
    1
but found (query options: "") :
    NULL
    1
```
This is because the query plan has separate scans for the first and third column families. The test passes if those scans resolve to the same node but fails if they resolve to different nodes.

### PR DESCRIPTION
**Pull request by solongordon**
The fakeSpanResolver, which is used for the fakedist* logic test
configs, was causing spurious test failures for tables with multiple
column families. The reason for this is that we sometimes generate
multiple spans for a point lookup if not all the column families are
needed. However, the fakeSpanResolver generates new splits and new
replica assignments on every call to Seek(). This means that spans for a
single row could get "split" across multiple replicas, which the SQL
layer is not equipped to deal with.

My somewhat hacky fix is to enforce that after Seek() is called, the
first range is assigned to the same replica as the last range from the
previous Seek() call. This ensures that if point lookups on the same
row are passed to Seek() in succession, they will not end up on
different replicas.

Fixes #44323

Release note: None

### CODE REVIEW

**Comment by rohany** [comment]
I know nothing about this part of the code, leaving it to radu

**Comment by RaduBerinde** [comment]
A bit worried that this restriction will reduce the space of possibilities too much. For example, now any two row (single col family) table will always map to a single range, right?

What would be a "correct" fix? We could remember the last key (minus any column family part) and see if is the same, would that be sufficient?

**Comment by solongordon** [comment]
Yes, that would be a more correct fix. For some reason I was under the impression it was non-trivial to check if two keys were describing the same row but now I see it's straightforward. I'll make that change.

**Comment by rohany** [comment]
I think you'll want to use `pkg/keys/keys.go:GetRowPrefixLength` to handle corner cases of metadata keys etc.

**Comment by solongordon** [comment]
RFAL assuming the logic tests don't blow up on TeamCity.

**Review by RaduBerinde — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44518)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @RaduBerinde, @rohany, and @solongordon)

___
*[pkg/sql/physicalplan/fake_span_resolver.go, line 144 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44518#-LzoVXsBAFE2xTueP_i0:-LzoVXsC7X9rjw01UPbR:bglra6k) ([raw file](https://github.com/cockroachdb/cockroach/blob/32322486a19a0b90b8d11564ec8d3943769df880/pkg/sql/physicalplan/fake_span_resolver.go#L144)):*
> ```Go
> 	}
> 
> 	// Check for the case where the previous Seek() call was for the same
> ```

Note that `Next()` modifies`fit.splits`. In a normal usage pattern, I'd expect `fit.splits` here to have 1 key, namely the last `span.EndKey` (or `span.Key` if descending). If that's the intention (it makes sense to me), it would be more clear if we used `fit.splits[len(fit.splits)-1]` or saved the last `fakeSplit` in a separate field.
___
*[pkg/sql/physicalplan/fake_span_resolver.go, line 148 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44518#-LzoWOZq3iltcjWTp7tv:-LzoWOZq3iltcjWTp7tw:b-e3odb8) ([raw file](https://github.com/cockroachdb/cockroach/blob/32322486a19a0b90b8d11564ec8d3943769df880/pkg/sql/physicalplan/fake_span_resolver.go#L148)):*
> ```Go
> 	// column families of the same row across different replicas.
> 	if fit.splits != nil {
> 		prefixLen, err := keys.GetRowPrefixLength(span.Key)
> ```

[nit] consider using `EnsureSafeSplitKey` for consistency (it's used above)
___
*[pkg/sql/physicalplan/fake_span_resolver.go, line 149 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44518#-LzoWSAFB7Ev2LEN_uvn:-LzoWSAFB7Ev2LEN_uvo:bjcquks) ([raw file](https://github.com/cockroachdb/cockroach/blob/32322486a19a0b90b8d11564ec8d3943769df880/pkg/sql/physicalplan/fake_span_resolver.go#L149)):*
> ```Go
> 	if fit.splits != nil {
> 		prefixLen, err := keys.GetRowPrefixLength(span.Key)
> 		if err == nil && len(fit.splits[0].key) >= prefixLen &&
> ```

[nit] add a comment explaining why it's ok to skip this in the error case (the error indicates that the start key can't possibly be "inside" a row).

**Review by solongordon — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44518)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @RaduBerinde, and @rohany)

___
*[pkg/sql/physicalplan/fake_span_resolver.go, line 144 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44518#-LzoVXsBAFE2xTueP_i0:-LzqrJsg45ji_5UvnwpU:bxmimfe) ([raw file](https://github.com/cockroachdb/cockroach/blob/32322486a19a0b90b8d11564ec8d3943769df880/pkg/sql/physicalplan/fake_span_resolver.go#L144)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Note that `Next()` modifies`fit.splits`. In a normal usage pattern, I'd expect `fit.splits` here to have 1 key, namely the last `span.EndKey` (or `span.Key` if descending). If that's the intention (it makes sense to me), it would be more clear if we used `fit.splits[len(fit.splits)-1]` or saved the last `fakeSplit` in a separate field.
</blockquote></details>

I think `fit.splits` actually has two elements here: the first with the start key and the second with the end key. And the replica we want to match is the one associated with the start key. I'll add a comment to that effect since it's confusing.
___
*[pkg/sql/physicalplan/fake_span_resolver.go, line 148 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44518#-LzoWOZq3iltcjWTp7tv:-LzqrAiw9pQItNCD3IMx:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/32322486a19a0b90b8d11564ec8d3943769df880/pkg/sql/physicalplan/fake_span_resolver.go#L148)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] consider using `EnsureSafeSplitKey` for consistency (it's used above)
</blockquote></details>

Done.
___
*[pkg/sql/physicalplan/fake_span_resolver.go, line 149 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44518#-LzoWSAFB7Ev2LEN_uvn:-Lzqu_se4quunSFJTvmg:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/32322486a19a0b90b8d11564ec8d3943769df880/pkg/sql/physicalplan/fake_span_resolver.go#L149)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] add a comment explaining why it's ok to skip this in the error case (the error indicates that the start key can't possibly be "inside" a row).
</blockquote></details>

Done.

**Review by andreimatei — COMMENTED** [review]
So you're saying `scanNode.spans` can end up having multiple spans for the same row?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44518#-:-Lzr6E4KCdMmS0rgqcvk:b-ylmiss)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @RaduBerinde, @rohany, and @solongordon)

___
*[pkg/sql/physicalplan/fake_span_resolver.go, line 149 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44518#-LzoWSAFB7Ev2LEN_uvn:-Lzr4sh30qyRRR2vEXAu:buyar1k) ([raw file](https://github.com/cockroachdb/cockroach/blob/32322486a19a0b90b8d11564ec8d3943769df880/pkg/sql/physicalplan/fake_span_resolver.go#L149)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

I don't understand this. I think `EnsureSafeSplitKey()` never returns errors for table spans. I think we should assert that there's no error and unindent all this. Or second best, set `fit.err`.
___
*[pkg/sql/physicalplan/fake_span_resolver.go, line 148 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44518#-Lzr5oFE5We2QBpPxNWQ:-Lzr5oFE5We2QBpPxNWR:babx1lt) ([raw file](https://github.com/cockroachdb/cockroach/blob/1be88207acfca9e323444c9e1835f64b97e57d81/pkg/sql/physicalplan/fake_span_resolver.go#L148)):*
> ```Go
> 	// column families of the same row across different replicas.
> 	if fit.splits != nil {
> 		// In the normal usage pattern, Next() is called until fit.splits has two
> ```

I can't really validate this comment, and even if I could it's hard to follow. Also the "normal usage pattern" thing scares me, I don't know what it means. I think it'd be much better to keep some information explicitly about the previous seek rather than relying on some side effects of iteration.

**Review by RaduBerinde — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44518)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RaduBerinde, @rohany, and @solongordon)

___
*[pkg/sql/physicalplan/fake_span_resolver.go, line 144 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44518#-LzoVXsBAFE2xTueP_i0:-Lzr6AgMCvItV_x8vCfL:b-4e857t) ([raw file](https://github.com/cockroachdb/cockroach/blob/32322486a19a0b90b8d11564ec8d3943769df880/pkg/sql/physicalplan/fake_span_resolver.go#L144)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

I think `fit.splits` actually has two elements here: the first with the start key and the second with the end key. And the replica we want to match is the one associated with the start key. I'll add a comment to that effect since it's confusing.
</blockquote></details>

Oh, this is pretty confusing. Why aren't we looking at the last key? We don't expect Seek calls to be overlapping. My hunch is that in the specific case we are addressing here, the start and end keys are in the same row so it's all the same, but I want to make sure there's some coherency here. Can you check if using `fit.splits[len(fit.splits)-1]` also fixes the issue?

**Review by solongordon — COMMENTED** [review]
> So you're saying scanNode.spans can end up having multiple spans for the same row?

Yes, for instance if a row has three column families and we only need to scan the first and last.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44518#-:-LzrPwsCDjZjZ7r3Ptiy:b-jlatb8)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @RaduBerinde, and @rohany)

___
*[pkg/sql/physicalplan/fake_span_resolver.go, line 144 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44518#-LzoVXsBAFE2xTueP_i0:-LzrEow_ADfI-tUhCxmO:b-l65m3p) ([raw file](https://github.com/cockroachdb/cockroach/blob/32322486a19a0b90b8d11564ec8d3943769df880/pkg/sql/physicalplan/fake_span_resolver.go#L144)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Oh, this is pretty confusing. Why aren't we looking at the last key? We don't expect Seek calls to be overlapping. My hunch is that in the specific case we are addressing here, the start and end keys are in the same row so it's all the same, but I want to make sure there's some coherency here. Can you check if using `fit.splits[len(fit.splits)-1]` also fixes the issue?
</blockquote></details>

We do look at the last key but only to use it as the end key for the final range. Its replica is never used.

I think the confusing thing here is that we store the replicas on the splits when really they are associated with the ranges between the splits. I'm refactored the existing code to make this more explicit. Let me know if you think this helps.
___
*[pkg/sql/physicalplan/fake_span_resolver.go, line 149 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44518#-LzoWSAFB7Ev2LEN_uvn:-LzrFWi12J1SznjkiKxW:bqv7j4p) ([raw file](https://github.com/cockroachdb/cockroach/blob/32322486a19a0b90b8d11564ec8d3943769df880/pkg/sql/physicalplan/fake_span_resolver.go#L149)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I don't understand this. I think `EnsureSafeSplitKey()` never returns errors for table spans. I think we should assert that there's no error and unindent all this. Or second best, set `fit.err`.
</blockquote></details>

Good to know. For some reason I thought I hit the error case here but now I'm not so sure. I changed it to set `fit.err` and we'll see if anything fails in CI.
___
*[pkg/sql/physicalplan/fake_span_resolver.go, line 148 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44518#-Lzr5oFE5We2QBpPxNWQ:-LzrFLWw0-4oxLPOhAhX:baxd9zz) ([raw file](https://github.com/cockroachdb/cockroach/blob/1be88207acfca9e323444c9e1835f64b97e57d81/pkg/sql/physicalplan/fake_span_resolver.go#L148)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I can't really validate this comment, and even if I could it's hard to follow. Also the "normal usage pattern" thing scares me, I don't know what it means. I think it'd be much better to keep some information explicitly about the previous seek rather than relying on some side effects of iteration.
</blockquote></details>

I did some refactoring to make things more explicit. Let me know if this helps.

**Review by RaduBerinde — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44518)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @rohany, and @solongordon)

___
*[pkg/sql/physicalplan/fake_span_resolver.go, line 144 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44518#-LzoVXsBAFE2xTueP_i0:-LzrWwfMCHf4Ya-xfGWt:buqmv4z) ([raw file](https://github.com/cockroachdb/cockroach/blob/32322486a19a0b90b8d11564ec8d3943769df880/pkg/sql/physicalplan/fake_span_resolver.go#L144)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

We do look at the last key but only to use it as the end key for the final range. Its replica is never used.

I think the confusing thing here is that we store the replicas on the splits when really they are associated with the ranges between the splits. I'm refactored the existing code to make this more explicit. Let me know if you think this helps.
</blockquote></details>

I understand. But the "normal" usecase for the resolver is where the spans passed to `Seek` are ordered and do not overlap (otherwise, this fix won't work reliably anyway). So from that standpoint, it makes more sense to look at the last end key. For example, say that we Seek for span `/1 - /2/fam1` (and say there are no splits) and then Seek for the span `/2/fam5 - /3`. We want to check that we are effectively skipping only part of a row; it shouldn't matter where the last span starts (the `/1`).

If you prefer to keep it simple and just address the case where the span passed to Seek is inside a single row, that's fine too but the code should then be restricted to only do something special in that case.

**Review by solongordon — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44518)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @RaduBerinde, and @rohany)

___
*[pkg/sql/physicalplan/fake_span_resolver.go, line 144 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44518#-LzoVXsBAFE2xTueP_i0:-LzrYrW6Bv4pUvwNQ31L:b3n5l2o) ([raw file](https://github.com/cockroachdb/cockroach/blob/32322486a19a0b90b8d11564ec8d3943769df880/pkg/sql/physicalplan/fake_span_resolver.go#L144)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I understand. But the "normal" usecase for the resolver is where the spans passed to `Seek` are ordered and do not overlap (otherwise, this fix won't work reliably anyway). So from that standpoint, it makes more sense to look at the last end key. For example, say that we Seek for span `/1 - /2/fam1` (and say there are no splits) and then Seek for the span `/2/fam5 - /3`. We want to check that we are effectively skipping only part of a row; it shouldn't matter where the last span starts (the `/1`).

If you prefer to keep it simple and just address the case where the span passed to Seek is inside a single row, that's fine too but the code should then be restricted to only do something special in that case.
</blockquote></details>

Oh, sorry I misunderstood your comment. Yes, looking at the last key makes sense to me. I'll change that.

**Review by solongordon — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44518)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @RaduBerinde, and @rohany)

___
*[pkg/sql/physicalplan/fake_span_resolver.go, line 149 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44518#-LzoWSAFB7Ev2LEN_uvn:-Lzr_Hhd-xDk4xcP0Dr3:bx2twx2) ([raw file](https://github.com/cockroachdb/cockroach/blob/32322486a19a0b90b8d11564ec8d3943769df880/pkg/sql/physicalplan/fake_span_resolver.go#L149)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

Good to know. For some reason I thought I hit the error case here but now I'm not so sure. I changed it to set `fit.err` and we'll see if anything fails in CI.
</blockquote></details>

It looks like `EnsureSafeSplitKey()` returns an error for keys which don't specify the column family, like `/Table/53/1/3`. I'm going to switch this back to swallowing the error unless you think there is a better solution.

**Review by RaduBerinde — APPROVED** [review]
:lgtm: Thanks for bearing with me :)

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44518#-:-LzrcNnq0iNXvxJkRqkx:bssyzpq)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andreimatei, @RaduBerinde, @rohany, and @solongordon)

___
*[pkg/sql/physicalplan/fake_span_resolver.go, line 164 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/44518#-Lzrc4ZK6JTJzPq63vRy:-Lzrc4ZK6JTJzPq63vRz:biq2kw) ([raw file](https://github.com/cockroachdb/cockroach/blob/870318f63c81625a2195a0be20184287771f0ea4/pkg/sql/physicalplan/fake_span_resolver.go#L164)):*
> ```Go
> 	// don't "split" column families of the same row across different replicas.
> 	if prevRange.endKey != nil {
> 		prefix, err := keys.EnsureSafeSplitKey(span.Key)
> ```

[nit] I'd invert the if so we don't have an early return, in case we add more code below in the future.

**Comment by solongordon** [comment]
TFTRs!

bors r+
